stan/math/prim/functor/closure_adapter.hpp

```diff
@@ -0,0 +1,104 @@
+#ifndef STAN_MATH_PRIM_FUNCTOR_CLOSURE_ADAPTER_HPP
+#define STAN_MATH_PRIM_FUNCTOR_CLOSURE_ADAPTER_HPP
+
+#include <stan/math/prim/meta/return_type.hpp>
+#include <ostream>
+
+namespace stan {
+namespace math {
+
+template <typename F>
+struct closure_adapter {
+  using captured_scalar_t__ = double;
+  using ValueOf__ = closure_adapter<F>;
+  static const size_t vars_count__ = 0;
+  F f_;
+
+  explicit closure_adapter(const F& f) : f_(f) {}
+
+  template <typename... Args>
+  auto operator()(std::ostream* msgs, Args... args) const {
+    return f_(args..., msgs);
+  }
+  auto value_of__() const { return closure_adapter<F>(f_); }
+  auto deep_copy_vars__() const { return closure_adapter<F>(f_); }
+  void zero_adjoints__() const {}
+  double* accumulate_adjoints__(double* dest) const { return dest; }
+  template <typename Vari>
+  Vari** save_varis(Vari** dest) const {
+    return dest;
+  }
+};
+
+template <typename F, typename T>
+struct simple_closure {
+  using captured_scalar_t__ = return_type_t<T>;
+  using ValueOf__ = simple_closure<F, decltype(value_of(std::declval<T>()))>;
+  const size_t vars_count__;
+  F f_;
+  T s_;
+
+  explicit simple_closure(const F& f, T s)
+      : f_(f), s_(s), vars_count__(count_vars(s)) {}
+
+  template <typename... Args>
+  auto operator()(std::ostream* msgs, Args... args) const {
+    return f_(s_, args..., msgs);
+  }
+  auto value_of__() const { return ValueOf__(f_, value_of(s_)); }
+  auto deep_copy_vars__() const {
+    return simple_closure<F, T>(f_, deep_copy_vars(s_));
+  }
+  void zero_adjoints__() { zero_adjoints(s_); }
+  double* accumulate_adjoints__(double* dest) const {
+    return accumulate_adjoints(dest, s_);
+  }
+  template <typename Vari>
+  Vari** save_varis__(Vari** dest) const {
+    return save_varis(dest, s_);
+  }
+};
+
+template <typename F>
+auto from_lambda(F f) {
+  return closure_adapter<F>(f);
+}
+
+template <typename F, typename T>
+auto from_lambda(F f, T a) {
+  return simple_closure<F, T>(f, a);
```

Comment:
Yes, that's right.

Actually, `from_lambda` isn't necessary because my plan was to leave implementing closure structs to the stanc codegen. `require_stan_closure_t<...>` recognizes closure simply by the presence of `::captured_scalar_t__`; they are effectively duck-typed. `simple_closure` is one example of how to make one.
Of course, an alternative (complementary?) path forward is expanding `simple_closure` so that the template type `T` is a parameter pack.

Other higher-order functions just need support for variadic arguments. Shortly after submitting this PR I realized you could revert all the changes to ODEs and instead have an adapter like this
```cpp
struct closure_adapter_ode {
  template<typename F, typename T0, typename T1, typename... Args>
  auto operator()(const T0& t, const Eigen::Matrix<T1, Eigen::Dynamic, 1>& y,
                  std::ostream* msgs, F& f, Args... args) {
    return f(msgs, t, y, args...);
  }
}

template <typename F, typename T_y0, typename T_t0, typename T_ts,
          typename... Args, require_stan_closure_t<F>* = nullptr>
std::vector<Eigen::Matrix<stan::return_type_t<T_y0, T_t0, T_ts, Args...>,
                          Eigen::Dynamic, 1>>
ode_rk45_tol_impl(const char* function_name, const F& f,
                  const Eigen::Matrix<T_y0, Eigen::Dynamic, 1>& y0_arg, T_t0 t0,
                  const std::vector<T_ts>& ts, double relative_tolerance,
                  double absolute_tolerance,
                  long int max_num_steps,
                  std::ostream* msgs, const Args&... args) {
  closure_adapter_ode adapter;
  return ode_rk45_tol_impl(function_name, adapter, y0_arg, t0, ts,
                           relative_tolerance, absolute_tolerance,
                           max_num_steps, msgs, f, args...);
}
```
Same goes for `reduce_sum` and others.